File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
# 프로그래머스 43105 정수 삼각형 LV.3
def solution(triangle):
    dp = [[0 for _ in i] for i in triangle]
    dp[0][0] = triangle[0][0]
    for i in range(len(triangle)):
        for j in range(i):
            dp[i][j] = max(dp[i][j], dp[i - 1][j] + triangle[i][j])
            dp[i][j + 1] = dp[i - 1][j] + triangle[i][j + 1]
    return max(dp[-1])

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;

class Solution {
    int[][] dp;
    
    public int solution(int[][] triangle) {
        int h = triangle.length;
        dp = new int[h][h];
        dp[0][0] = triangle[0][0];
        
        for (int i = 1; i < h; i++) {
            for (int j = 0; j < i; j++) {
                dp[i][j] = Math.max(dp[i][j], dp[i - 1][j] + triangle[i][j]);
                dp[i][j + 1] = Math.max(dp[i][j + 1], dp[i - 1][j] + triangle[i][j + 1]);
            }
        }
        
        int max = Integer.MIN_VALUE;
        for (int x : dp[h - 1]) {
            if (max < x) max = x;
        }
        
        return max;
    }
}